x-pack/metricbeat/module/oracle/performance/data.go

```diff
@@ -0,0 +1,138 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package performance
+
+import (
+	"context"
+
+	"github.com/elastic/beats/x-pack/metricbeat/module/oracle"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+// extract is the E of a ETL processing. Gets the data files, used/free space and temp free space data that is fetch
+// by doing queries to Oracle
+func (m *MetricSet) extract(ctx context.Context, extractor performanceExtractMethods) (out *extractedData, err error) {
+	out = &extractedData{}
+
+	if out.bufferCacheHitRatios, err = extractor.bufferCacheHitRatio(ctx); err != nil {
+		return nil, errors.Wrap(err, "error getting buffer cache hit ratio")
+	}
+
+	if out.libraryData, err = extractor.library(ctx); err != nil {
+		return nil, errors.Wrap(err, "error getting library data")
+	}
+
+	if out.cursorsByUsernameAndMachine, err = extractor.cursorsByUsernameAndMachine(ctx); err != nil {
+		return nil, errors.Wrap(err, "error getting cursors by username and machine")
+	}
+
+	if out.totalCursors, err = extractor.totalCursors(ctx); err != nil {
+		return nil, errors.Wrap(err, "error getting total cursors")
+	}
+
+	return
+}
+
+func (m *MetricSet) extractAndTransform(ctx context.Context) ([]mb.Event, error) {
+	extractedMetricsData, err := m.extract(ctx, m.extractor)
+	if err != nil {
+		return nil, errors.Wrap(err, "error extracting data")
+	}
+
+	return m.transform(extractedMetricsData), nil
+}
+
+// transform is the T of an ETL (refer to the 'extract' method above if you need to see the origin). Transforms the data
+// to create a Kibana/Elasticsearch friendly JSON. Data from Oracle is pretty fragmented by design so a lot of data
+// was necessary. Data is organized by Tablespace entity (Tablespaces might contain one or more data files)
+func (m *MetricSet) transform(in *extractedData) []mb.Event {
+	bufferCache := m.addBufferCacheRatioData(in.bufferCacheHitRatios)
+	cursorByUsernameAndMachineEvents := m.addCursorByUsernameAndMachine(in.cursorsByUsernameAndMachine)
+
+	cursorEvent := m.addCursorData(in.totalCursors)
+	cursorEvent.Update(m.addLibraryData(in.libraryData))
+
+	events := make([]mb.Event, 0)
+
+	events = append(events, mb.Event{MetricSetFields: cursorEvent})
+
+	for _, v := range cursorByUsernameAndMachineEvents {
+		events = append(events, mb.Event{MetricSetFields: v})
+	}
+
+	for _, v := range bufferCache {
+		events = append(events, mb.Event{MetricSetFields: v})
+	}
```

Comment:
It is not so clear to me what events this metricset generates, From what I understand it is generating for each fetch/transform:
* An event with aggregated cursor data, I guess that this the one in the example?
* An event with cursor data per username and machine. It seems that containing the username and machine.
* Several buffer cache events, why are cache stats split in several events?

I wonder if the data per cursor should be in a different metricset, as I guess it will generate loads on events if there are loads of connections.

It'd be good if you could add some explanation about the events generated to the documentation, or to the godocs of these methods.

You can also collect multiple data files as is done [in the socket metricset](https://github.com/elastic/beats/tree/master/metricbeat/module/system/socket/_meta). I think it can be useful in this module.

Reply:
I agree it wasn't clear, but I didn't know you we can write several events in `_meta` folder so I have added all.

I see your point in splitting events but the number of potential Metricsets were way too many while they would be related between them. At first I thought about splitting cursor and cache metricsets but then cursors also would need another subsplit or even more.

It was better to have a single Metricset that plays well with a single dashboard that to have dashboards which requires `a` and `b` metricsets, while other requires `b` and `c` even when both are performance dashboards. I don't know if I explained it well :smile: 

At the end I think that more "human" semantics in the metricsets are better that their direct "under the hood" meanings. So `tablespace` metricset was better than `data_file` and `temp_file` metricsets and `performance` metricset is better than `cursors`, `library_cache`, `cache` `io`... `activity` metricset is also better than `sessions`, `process` blah blah... just examples, don't take it literally

Reply:
Ok, in any case if they are too many metrics users can always drop with processors.